Timezone aware scale

paulrosenzweig opened the pull request:
This PR extracts the core timezone-aware range from #7247 and fixes a bug.

paulrosenzweig committed:
copy over code from pr

paulrosenzweig committed:
remove chronological

paulrosenzweig committed:
remove tick formatting

paulrosenzweig committed:
remove chronological

paulrosenzweig committed:
use reporting_timezone

paulrosenzweig committed:
remove unneeded tests

paulrosenzweig committed:
remove import

paulrosenzweig committed:
add timeseriesScale tests/fixes

paulrosenzweig committed:
eslint

paulrosenzweig committed:
more timeseriesScale tests

paulrosenzweig committed:
remove now unneeded rangeFn

paulrosenzweig committed:
more timeseriesScale tests

paulrosenzweig committed:
report_timezone to actual_timezone

paulrosenzweig committed:
display warning if timezones are mismatched

paulrosenzweig committed:
extract getTimezone

paulrosenzweig committed:
don't error if series doesn't have a card (in tests)

tlrobinson commented on frontend/src/metabase/query_builder/selectors.js:
```diff
@@ -244,15 +244,21 @@ export const getRawSeries = createSelector(
+            display: display,
+            visualization_settings: settings,
+            dataset_query: lastRunDatasetQuery,
+            expected_timezone,
```
I'm strong 👎 on copying the timezone to the `card` definition, or really requiring the frontend copy it from the response to anywhere else in `card`/`data`/`series`. This is only one of several places where queries are run (e.x. dashboards, QuestionAndResultLoader, probably others). I haven't tried it but it's unlikely this will work correctly anywhere outside of the query builder.

`Visualization` component takes the `series` array with objects containing `card` definition (a.k.a. Question, primarily for `display` and `visualization_settings`) and a `data` result set. We shouldn't require any processing of `card` and `data` obtained from the backend (aside from zippering them into `series`). The only reason we do in the code above is because we allow the user to modify the `display` and `visualization_settings` in the query builder.

This is also one of the reasons I was advocating for putting the timezone info on the column itself: even if it's always going to be the same across all columns in the query, we *use* that information on a column level, so it's easiest to ensure it's passed all the way down where it's used by including it on the column. I'd also be ok with it being in the `data` object though.

Actually, now that I think about it, with dashboard multiseries there's a possibility you could try to display two queries that were run in different timezones. We'd probably want to just pick the first timezone and show a warning.  See comments in the code below.

tlrobinson commented on frontend/src/metabase/visualizations/lib/apply_axis.js:
```diff
@@ -160,16 +154,13 @@ export function applyChartTimeseriesXAxis(
 
   // set the x scale
-  chart.x(d3.time.scale.utc().domain(xDomain)); //.nice(d3.time[dataInterval.interval]));
+  chart.x(timeseriesScale(tickInterval, getTimezone(series)).domain(xDomain));
```
As mentioned above, there's a chance you could have multiple series with different timezones (via dashboard multiseries). Here's what I'd do:

Change `getTimezone` to take a single series object, depending on where `actual_timezone` is moved to:

```
function getTimezone({ data }) {
  // if we put it on the column:
  return data.cols[0].actual_timezone;
  // if we put it on data:
  return data.actual_timezone;
}
```

In this function, do something like:

```
const timezones = _.uniq(series.map(s => getTimezone(s)))
if (timezones.length > 1) {
  warn(multipleTimezoneWarning({ timezones }))
}
const timezone = timezones[0] || DEFAULT_TIMEZONE;

chart.x(timeseriesScale(tickInterval, timezone).domain(xDomain));
```

Not sure if we need a `DEFAULT_TIMEZONE` but maybe just to be safe?

paulrosenzweig committed:
move expected/actual timezones into data, add warning when series have different timezones

paulrosenzweig committed:
fixes

paulrosenzweig commented:
@tlrobinson Good call! I updated this to nest the timezone fields in data. I also added another warning to cover dashboard multiseries with multiple values for `actual_timezone`.

@camsaul how badly did I massacre the Clojure? Is `add_settings` now misnamed?

paulrosenzweig committed:
make sure getTimezone works on transformed series

paulrosenzweig committed:
update actual/expected to results/requested

paulrosenzweig committed:
call getTimezone from LAB renderer rather than apply axis

paulrosenzweig committed:
include timezone as part of a timeseries interval

paulrosenzweig committed:
missed variable

paulrosenzweig committed:
include timezone when updating timeseries interval

tlrobinson commented on frontend/src/metabase/visualizations/lib/apply_axis.js:
```diff
@@ -135,20 +133,15 @@ export function applyChartTimeseriesXAxis(
     chart.xAxis().tickFormat(timestamp => {
-      // timestamp is a plain Date object which discards the timezone,
-      // so add it back in so it's formatted correctly
-      const timestampFixed = moment(timestamp)
```
yay!

tlrobinson commented on frontend/test/metabase/visualizations/components/LineAreaBarRenderer.unit.spec.js:
```diff
@@ -166,7 +124,12 @@ describe("LineAreaBarRenderer", () => {
+    const series = [
+      {
+        data: { cols, rows },
+        card: { display: chartType, report_timezone: "Etc/UTC" },
```
`report_timezone` was removed right?

paulrosenzweig replied:
Good catch! It was working anyways because "Etc/UTC" is the default anyways. I just deleted it here.

tlrobinson reviewed (approved):
Looks good, just one minor thing in a test and it would be nice to add `requested_timezone` to the Flow type for `DatasetData` even though they're apparently not used in any of this code (`frontend/src/metabase/meta/types/Dataset.js`)

paulrosenzweig committed:
remove "report_timezone"

paulrosenzweig committed:
add timezone fields to DatasetData flow type

paulrosenzweig committed:
make the fields optional

camsaul reviewed (approved):
works for me

paulrosenzweig committed:
Fill data using timezone scale (#11143)
